### PR TITLE
Hardening type evolution.

### PR DESCRIPTION
**Pull request by dkorolev**
(Note: This PR contains all the changes from https://github.com/C5T/Current/pull/528).

Hi @mzhurovich , CC @grixa.

**TL;DR:**

1. The implementation of our type evolution is officially robust to the degree it can handle CTFO's ID migration. Specifically, I've added and tested `CURRENT_ENUM`, `std::pair<>`, and `Optional<>` support.
2. Sherlock streams, and, by extension, our Storage subsystem, not only expose all their types in a format ready for type evolution, but also "create" a `CURRENT_NAMESPACE` (actually, a `CURRENT_DERIVED_NAMESPACE`) of a user-defined name, exposing the `TopLevelTransaction` type. (*) Thus, as long as while initializing the `Storage` from the user code (CTFO++) we keep this namespace different from version to version, most of data&schema migration would would be about `#include`-ing that one `curl`-ed "header".

But of course:

```
[----------] Global test environment tear-down
[==========] 376 tests from 57 test cases ran. (5857 ms total)
[  PASSED  ] 376 tests.
```

PTAL && thanks!
Dima

(*) I've realized too late `TopLevelTransaction` makes little sense for the Stream itself, it's a Storage-centric name. Accepting suggestions.

### CODE REVIEW

File: Sherlock/sherlock.h [hunk 1]
```diff
@@ -483,7 +483,10 @@ class StreamImpl {
     explicit FillPerLanguageSchema(SherlockSchema& schema) : schema_ref(schema) {}
     template <current::reflection::Language language>
     void PerLanguage() {
-      schema_ref.language[current::ToString(language)] = schema_ref.type_schema.Describe<language>();
+      current::reflection::NamespaceToExpose expose("Sherlock");
+      expose.template AddType<entry_t>("TopLevelTransaction");
```
Comment (mzhurovich) [inline Sherlock/sherlock.h]: `StreamEntry(Type)`?

**Comment by mzhurovich** [comment]
@dkorolev I doubt, I can do something more constructive at this point than just merge this pull request :) This is really huge step and I really appreciate your effort to convert our ideas into this solid prototype. Так что, помолясь, применимъ!

**Comment by dkorolev** [comment]
Yay! Thanks @mzhurovich 

@grixa -- This is a pretty big change, perhaps now is a good time to re-clone a fresh repo altogether on your end, confirm the top-level `make test` and `make check` pass, and take it from there? Thanks!
